pkg/cmd/server/start/bootstrap_node.go

```diff
@@ -0,0 +1,339 @@
+	}
+
+	kubeClientConfig, err := o.NodeArgs.KubeConnectionArgs.ClientConfig.ClientConfig()
+	if err != nil {
+		return err
+	}
+	var c kclientset.Interface
+	c, err = kclientset.NewForConfig(kubeClientConfig)
+	if err != nil {
+		return err
+	}
+
+	glog.Infof("Bootstrapping from API server %s (experimental)", kubeClientConfig.Host)
+
+	c, err = o.loadBootstrapClientCertificate(nodeConfigDir, c, 1*time.Minute)
+	if err != nil {
+		return err
+	}
+	if err := o.loadBootstrapServerCertificate(nodeConfigDir, hostnames, c, 1*time.Minute); err != nil {
+		return err
+	}
+
+	nodeClientCAPath := filepath.Join(nodeConfigDir, "node-client-ca.crt")
+	if len(kubeClientConfig.CAData) > 0 {
+		if err := ioutil.WriteFile(nodeClientCAPath, []byte(kubeClientConfig.CAData), 0600); err != nil {
+			return err
+		}
+	}
+
+	// try to refresh the latest node-config.yaml
+	o.ConfigFile = filepath.Join(nodeConfigDir, "node-config.yaml")
+	config, err := c.Core().ConfigMaps("openshift-infra").Get("node-config")
+	if err == nil {
+		// skip all the config we generated ourselves
+		skipConfig := map[string]struct{}{"server.crt": {}, "server.key": {}, "master-client.crt": {}, "master-client.key": {}, "node.kubeconfig": {}}
+		for k, v := range config.Data {
+			if _, ok := skipConfig[k]; ok {
+				continue
+			}
+			b := []byte(v)
+			// if a node config is provided, override the setup.
+			if k == "node-config.yaml" {
+				if err := ioutil.WriteFile(o.ConfigFile, []byte(v), 0600); err != nil {
+					return err
+				}
+				nodeConfig, err := configapilatest.ReadNodeConfig(o.ConfigFile)
+				if err != nil {
+					return err
+				}
+				if err := o.NodeArgs.MergeSerializeableNodeConfig(nodeConfig); err != nil {
+					return err
+				}
+				nodeConfig.ServingInfo.ServerCert.CertFile = filepath.Join(nodeConfigDir, "server.crt")
+				nodeConfig.ServingInfo.ServerCert.KeyFile = filepath.Join(nodeConfigDir, "server.key")
+				nodeConfig.ServingInfo.ClientCA = nodeClientCAPath
+				nodeConfig.MasterKubeConfig = filepath.Join(nodeConfigDir, "node.kubeconfig")
+				b, err = configapilatest.WriteYAML(nodeConfig)
+				if err != nil {
+					return err
+				}
+			}
+
+			if err := ioutil.WriteFile(filepath.Join(nodeConfigDir, k), b, 0600); err != nil {
+				return err
+			}
+		}
+		glog.V(3).Infof("Received %d bootstrap files into %s", len(config.Data), nodeConfigDir)
+	}
+
+	// if we had a previous node-config.yaml, continue using it
+	if _, err2 := os.Stat(o.ConfigFile); err2 == nil {
+		if err == nil {
+			glog.V(2).Infof("Unable to load node configuration from the server: %v", err)
+		}
+		return nil
+	}
+
+	// if there is no node-config.yaml and no server config map, generate one
+	if kerrors.IsNotFound(err) {
+		glog.V(2).Infof("Generating a local configuration since no server config available")
+		nodeConfig, err := o.NodeArgs.BuildSerializeableNodeConfig()
+		if err != nil {
+			return err
+		}
+		if err := o.NodeArgs.MergeSerializeableNodeConfig(nodeConfig); err != nil {
```

Comment:
have you sanity-checked if there's any difference between this and `start node --write-config`?